src/babashka/process.cljc

```diff
@@ -413,8 +413,16 @@
                          prev
                          cmd)]
       (when shutdown
-        (-> (Runtime/getRuntime)
-            (.addShutdownHook (Thread. (fn [] (shutdown res))))))
+        (let [hook (Thread. (fn [] (shutdown res)))]
+          (-> (Runtime/getRuntime)
+              (.addShutdownHook hook))
+          (if-before-jdk8
+           nil ;; throwing an exception here would commit breakage, therefore accept the memory leak for jdk8
+           (-> (.onExit proc)
```

Comment:
What if there is also a `:exit-fn`?

Reply:
Ah, I tested this, it works because you can call `.onExit` + `.thenRun` multiple times.

```
user=> (require '[babashka.process :as p])
nil
user=> (let [p (:proc (p/process "sleep" "1"))] (-> (.onExit p) (.thenRun (fn [] (prn :foo)))) (-> (.onExit p) (.thenRun (fn [] (prn :bar)))))
#object[java.util.concurrent.CompletableFuture 0x77049094 "java.util.concurrent.CompletableFuture@77049094[Not completed]"]
user=> :bar
:foo
```

Reply:
@maxweber Can I ask you what kind of stuff you were doing in the shutdown hook which caused the memory build-up?

Reply:
I think the only downside to this PR is that the shutdown hook won't be executed if the process finishes earlier than the exit of the parent process, but I guess this is fine since the shutdown hook is mostly there for killing the process and its children.

Reply:
Hmm, one corner case: if you previously had `:shutdown p/destroy-tree` then the PR behavior might result in not having the children killed?

Reply:
@borkdude we have a server that polls for log entries from three other servers. This is done every second for each server. 

Regarding your last question: maybe calling the shutdown hook / starting the thread before removing it would cover the mentioned corner case?

Reply:
That seems to just work. So adding `(shutdown res)` after removing the shutdown hook would work I think to ensure it's still called.

Reply:
@borkdude I'm using `p/destroy(-tree)` indirectly via the defaults of `process/shell`. In our code I just set `:shutdown nil` to fix the memory leak for the moment :)

Adding `(shutdown res)` after the hook removal sounds good 👍 One thought, maybe starting the hook Thread instead would be safer? If someone does something 'fancy' in the hook that blocks very long or just has an error that causes an exception usually only during the shutdown of a long running app?

Reply:
@borkdude as discussed the `:shutdown` hook is now executed as soon as the child process ends and the docstring has been extended.